### DISCUSSION TITLE
Logical search on tags

### DISCUSSION THREAD

**Discussion by Erwyn** [kimai/kimai#3063] [Questions & Answers]
Hi there,

I tried searching in the discussions, if it's already answered feel free to just point me to the right place.

One thing I'd like to do is to be able to search on multiple tags using some logical operators. Let's say I have four entries with tags:

1. dev, php
2. qa, php 
3. dev, ruby
4. dev

I want to select:

1. everything "dev". This is already possible by just searching "dev"
2. entry 1, so everything "dev" AND "php"
3. entry 3, everything "dev" BUT NOT "php"
4. entry 4, everything ONLY "dev"

`1.` is the default behaviour so this is fine, but I don't know how to achieve `2.` , `3.` and `4.`. Is there a way to do some advanced search based on tags I'm not aware of ?

**Comment by kevinpapst** [comment]
The search in general does not support negation only `AND`.
It sounds as if you want to fid timesheets, is that correct?
The tag search afair depends on the mode that is active. If you have a dropdown, it will do exact matches on the chosen tags.
If you have "freestyle" tags, so a blank text fields, it might behave a bit different.
Can you please be more specific including your system setting:
<img width="476" alt="Bildschirmfoto 2022-01-10 um 12 08 17" src="https://user-images.githubusercontent.com/533162/148756376-e50c976c-1316-4e7e-8ac0-11e43e6285db.png">

**Reply by Erwyn** [reply]
So, what I'm trying to do is to clear (mark as exported) records, so in the Export interface, but I wasn't able to make it work in the "My times" interface either. For me it's always doing a OR operation if I specify multiple tags to be searched.

![image](https://user-images.githubusercontent.com/206743/148759565-7e1f21c6-0673-488d-835c-a0e291ae9621.png)

**Reply by kevinpapst** [reply]
OR is not implemented, so I don't see how that should work:
https://github.com/kevinpapst/kimai2/blob/master/src/Repository/TimesheetRepository.php#L998

Try to change the tag mode and search again.

**Reply by Erwyn** [reply]
So, I changed the configuration:

![image](https://user-images.githubusercontent.com/206743/148763589-f8f0e281-acc8-41d8-9c16-510dec6b06b7.png)

and here is what I search for:

![image](https://user-images.githubusercontent.com/206743/148764064-e585dc3b-3097-4e9a-9edb-d16c37473cc7.png)

and here is what I get:

![image](https://user-images.githubusercontent.com/206743/148764184-f74f40d3-086f-46b1-b5af-4e65e82ac14b.png)

**Reply by kevinpapst** [reply]
Aha, yeah so both my AND and your OR assumption is not correct. 
It just fetches every record that has at least one of the selected tags.

If you have an idea how to improve that: please share it! I would also gladly review a PR if you want to work on a solution.

**Reply by Erwyn** [reply]
>  It just fetches every record that has at least one of the selected tags.

I think that is what a OR behaviour is. Just to be perfectly clear, I'm not trying to lecture you on the OR or whatever I couldn't care less :smiley:, there is no ego involved. It's just that in my head it's a OR behaviour, and if it's definitely not, then there is something I don't understand or don't see and I would love for you to enlight me ! 

Now, I completely understand that although it might (if I got it correctly) be a OR logically, that was not the intent programmatically. I could have a try at fixing that but I'm a bit afraid it might break other people's workflow if they depend on it being processed as an OR. Should it have been a AND since the beginning, is it really a bug ? If you decide so, I could give it a try.

**Reply by kevinpapst** [reply]
I believe Kimai shows that I do understand the difference between AND and OR - I just guess my programming skills might be better than my GitHub discussion skills 😁 sometimes explaining the thought in your head is complicated and takes too much resources, so you try to cut the corner by being slightly incorrect. My comment was targeted against the `isMemberOf` function which I believe compiles to an EXISTS(SUBSELECT) query. So from the pure SQL pov it is neither AND nor OR. But I believe this part of the discussion should stop right here, because it doesn't help.

What is important to me:
The search should behave the same in all places. 
If you select two customer, you want to see the record for both customer.
If you now change the behavior of the tags field to be an exclusive AND would mean: the tags field as only field in Kimai would lead to another kind of result. 
I believe if you want to fix that properly, you have to fix it from top to bottom by first coming up with a UI element, that allows to distinguish between AND and OR. That could be an expert mode that allows to type in the search "dev AND (php OR bash)" or something completely different.
